init/tee/kbs/kbs_curl.c

```diff
@@ -0,0 +1,164 @@
+        if (code != CURLE_OK)
+                return KBS_CURL_ERR("CURLOPT_WRITEFUNCTION");
+
+        /*
+         * Based on the operation TYPE, there will need to be some additional
+         * cURL attributes set. These will also perform the cURL request.
+         */
+        switch (type) {
+        case KBS_CURL_REQ:
+                return kbs_curl_post_request(curl, url, in, out);
+        default:
+                return KBS_CURL_ERR("Type argument invalid");
+        }
+}
+
+/*
+ * Set the cURL headers. If the session args is not NULL, that indicates that
+ * the session ID has been retrieved from attestation server before, and that
+ * session ID should be included in the headers.
+ */
+static CURLcode
+kbs_curl_set_headers(CURL *curl, char *session)
+{
+        struct curl_slist *slist;
+        char session_buf[100];
+
+        slist = NULL;
+        slist = curl_slist_append(slist, "Accept: application/json");
+        slist = curl_slist_append(slist,
+                "Content-Type: application/json; charset=utf-8");
+
+        /*
+         * Add the session ID cookie if the session ID exists.
+         */
+        if (session) {
+                sprintf(session_buf, "Cookie: session_id=%s", session);
+                curl_slist_append(slist, session_buf);
+        }
+
+        /*
+         * Set the headers.
+         */
+        return curl_easy_setopt(curl, CURLOPT_HTTPHEADER, slist);
+}
+
+/*
+ * POST a KBS REQUEST and attempt to retrieve a nonce from the attestation
+ * server.
+ */
+static int
+kbs_curl_post_request(CURL *curl, char *url, char *req, char *nonce)
+{
+        CURLcode code;
+        char req_url[100];
+
+        /*
+         * Set the cURL POST URL to $URL/kbs/v0/auth.
+         */
+        sprintf(req_url, "%s/kbs/v0/auth", url);
+
+        code = curl_easy_setopt(curl, CURLOPT_URL, req_url);
+        if (code != CURLE_OK)
+                return KBS_CURL_ERR("CURLOPT_URL");
+
+        /*
+         * Set the KBS REQUEST size and data.
+         */
+        code = curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, (long) strlen(req));
+        if (code != CURLE_OK)
+                return KBS_CURL_ERR("CURLOPT_POSTFIELDSIZE");
+
+        code = curl_easy_setopt(curl, CURLOPT_POSTFIELDS, req);
+        if (code != CURLE_OK)
+                return KBS_CURL_ERR("CURLOPT_POSTFIELDS");
+
+        /*
+         * We would like the nonce written to the "nonce" argument.
+         */
+        code = curl_easy_setopt(curl, CURLOPT_WRITEDATA, nonce);
+        if (code != CURLE_OK)
+                return KBS_CURL_ERR("CURLOPT_WRITEDATA");
+
+        code = curl_easy_perform(curl);
+        if (code != CURLE_OK && code != CURLE_WRITE_ERROR)
+                return KBS_CURL_ERR("CURL_EASY_PERFORM");
+
+        return 0;
+}
+
+/*
+ * Based on the given cURL error, print the KBS error and return an error
+ * indicator.
+ */
+static int
+KBS_CURL_ERR(char *errmsg)
+{
+        printf("ERROR (kbs_curl_post): %s\n", errmsg);
```

Comment:
I'm not entirely sure this function is useful. We're losing the ability to pass format strings (varargs) and it's used from more places than `kbs_curl_post`, so the function name is confusing.

Probably a macro that uses `__func__` to print the function name and supports passing varargs to `printf` would work better.

Reply:
Fixed in [28e1566](https://github.com/containers/libkrun/pull/108/commits/28e1566dca9ce33a7b1ec05fe3a2864fb7d64fce)